[DataLayout] Introduce sentinel pointer value

shiltian opened the pull request:
The value of a null pointer is not always `0`. For example, on AMDGPU, the null
pointer in address spaces 3 and 5 is `0xffffffff`. Currently, there is no
target-independent way to get this information, making it difficult and
error-prone to handle null pointers in target-agnostic code.

We do have `ConstantPointerNull`, but it might be a little confusing and
misleading. It represents a pointer with an all-zero value rather than
necessarily a real `nullptr`.

This PR introduces the concept of a *sentinel pointer value* to `DataLayout`,
representing the actual `nullptr` value for a given address space. The changes
include:

- A new interface function:
  ```
  APInt getSentinelPointerValue(unsigned AS)
  ```
  This function returns an `APInt` representing the sentinel pointer value for
  the given address space `AS`. An `APInt` is used instead of a literal integer
  to support cases where pointers are wider than 64 bits (e.g., AMDGPU’s address
  space 8).

- An extension to the data layout string format:
  ```
  p[n]:<size>:<abi>[:<pref>[:<idx>[:<sentinel>]]]
  ```
  The new `<sentinel>` component specifies the sentinel value for the
  corresponding pointer. It currently supports two values:
    - `z` for an all-zero value
    - `f` for a full-bit set value

  These two values are the most common representations of `nullptr`. It is
  unlikely that any target would define `nullptr` as a random value.

A follow-up patch series will introduce an equivalent of `ConstantPointerNull`
that represents the actual `nullptr`, built on top of this PR.

shiltian commented:
* **#131557** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/131557?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/131557?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shiltian commented:
The RFC is posted https://discourse.llvm.org/t/rfc-introduce-sentinel-pointer-value-to-datalayout/85265.

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -3143,7 +3143,10 @@ as follows:
     are in bits. The address space, ``n``, is optional, and if not specified,
     denotes the default address space 0. The value of ``n`` must be
-    in the range [1,2^24).
+    in the range [1,2^24). The fifth parameter ``<sentinel>`` specifies the
+    sentinel value of the pointer for the corresponding address space. It
+    currently accepts two values: ``0`` for an all-zero value and ``f`` for a
+    full-bit set value. The default sentinel pointer value is all-zero.
```
"no sentinel pointer" should also be a valid option. We also should be able to specify the default for unlisted address spaces (I thought the old PR had that part implemented?)

I'm also not sure if we should just allow an arbitrary bit pattern here.

This also should define what the sentinel pointer means. Also clarify how this interacts with null_pointer_is_valid and the various flavors of nonnull

arsenm reviewed (commented):
This needs an RFC. For reference a previous attempt was at #83109

arsenm commented on clang/lib/Basic/Targets/AMDGPU.cpp:
```diff
@@ -32,9 +32,9 @@ static const char *const DataLayoutStringR600 =
-    "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32"
-    "-p7:160:256:256:32-p8:128:128-p9:192:256:256:32-i64:64-v16:16-v24:32-v32:"
-    "32-v48:64-v96:128"
+    "e-p:64:64-p1:64:64-p2:32:32-p3:32:32:32:32:f-p4:64:64-p5:32:32:32:32:f"
```
p2 should also be -1

shiltian replied:
Our document says not implemented for HSA though.

arsenm replied:
Doesn't matter what HSA says about it. Graphics uses it and it exists

arsenm commented on clang/lib/Basic/Targets/AMDGPU.cpp:
```diff
@@ -32,9 +32,9 @@ static const char *const DataLayoutStringR600 =
-    "32-v48:64-v96:128"
+    "e-p:64:64-p1:64:64-p2:32:32-p3:32:32:32:32:f-p4:64:64-p5:32:32:32:32:f"
+    "-p6:32:32-p7:160:256:256:32-p8:128:128-p9:192:256:256:32-i64:64-v16:16"
+    "-v24:32-v32:32-v48:64-v96:128"
```
Not sure about the interaction with fat pointers. It only applies to a subset of the bits. But also not that important, we probably can't do much with that information

shiltian replied:
Right, that's why I didn't try to add that information to the fat pointer at the moment, and it is all-zero anyway.

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -13045,9 +13048,9 @@ This instruction requires several arguments:
    -  The call is in tail position (ret immediately follows call and ret
       uses value of call or is void).
-   -  Option ``-tailcallopt`` is enabled, ``llvm::GuaranteedTailCallOpt`` is 
+   -  Option ``-tailcallopt`` is enabled, ``llvm::GuaranteedTailCallOpt`` is
```
Unrelated whitespace changes

shiltian commented:
> This needs an RFC. For reference a previous attempt was at #83109

The RFC was posted, as mentioned in a previous comment.

shiltian commented on llvm/include/llvm/IR/DataLayout.h:
```diff
@@ -552,6 +553,11 @@ class DataLayout {
+  /// Returns the sentinel pointer value for a given address space. If the
+  /// address space is invalid, it defaults to the sentinel pointer value of
+  /// address space 0, aligning with the behavior of \p getPointerSpec.
+  APInt getSentinelPointerValue(unsigned AS) const;
```
I'd prefer to leave some space for future extension. The challenging part is how to "encode" it. Do we really want to allow something like `0xfffdddfffddd` in the data layout string?

shiltian committed:
[DataLayout] Introduce sentinel pointer value

The value of a null pointer is not always `0`. For example, on AMDGPU, the null
pointer in address spaces 3 and 5 is `0xffffffff`. Currently, there is no
target-independent way to get this information, making it difficult and
error-prone to handle null pointers in target-agnostic code.

We do have `ConstantPointerNull`, but it might be a little confusing and
misleading. It represents a pointer with an all-zero value rather than
necessarily a real `nullptr`.

This PR introduces the concept of a *sentinel pointer value* to `DataLayout`,
representing the actual `nullptr` value for a given address space. The changes
include:

- A new interface function:
  ```
  APInt getSentinelPointerValue(unsigned AS)
  ```
  This function returns an `APInt` representing the sentinel pointer value for
  the given address space `AS`. An `APInt` is used instead of a literal integer
  to support cases where pointers are wider than 64 bits (e.g., AMDGPU’s address
  space 8).

- An extension to the data layout string format:
  ```
  p[n]:<size>:<abi>[:<pref>[:<idx>[:<sentinel>]]]
  ```
  The new `<sentinel>` component specifies the sentinel value for the
  corresponding pointer. It currently supports two values:
    - `0` for an all-zero value
    - `f` for a full-bit set value

  These two values are the most common representations of `nullptr`. It is
  unlikely that any target would define `nullptr` as a random value.

A follow-up patch series will introduce an equivalent of `ConstantPointerNull`
that represents the actual `nullptr`, built on top of this PR.

arichardson commented on llvm/docs/LangRef.rst:
```diff
@@ -3143,7 +3143,10 @@ as follows:
-    in the range [1,2^24).
+    in the range [1,2^24). The fifth parameter ``<sentinel>`` specifies the
+    sentinel value of the pointer for the corresponding address space. It
+    currently accepts two values: ``0`` for an all-zero value and ``f`` for a
```
It would be nice if we could just use a normal number `0`/`-1` but since `-` is the separator that does work and having to spell out the full numeric all-ones value is even more ugly. While I get the rationale for `f` it is a bit odd having 0 be the other one. I'd say either go for letters in both cases `f`(ull)/`z`(ero) or `a`(ll)/`z`ero?

Another option could be `0` and `~0` for all ones?

shiltian committed:
Replace '0' with 'z' for all-zero value

shiltian committed:
Fix lld test failures
